File: dsa/li18_string/java/L5_LongestPalindromicSubstring.java

public class L5_LongestPalindromicSubstring {

	public static void main(String[] args) {
		L5_LongestPalindromicSubstring obj = new L5_LongestPalindromicSubstring();

		String s = "babad";
		// String s = "aacabdkacaa";
		// String s = "bb";
		System.out.println(obj.longestPalindrome(s));
		System.out.println(obj.longestPalindrome2(s));
	}
	// Problem: https://leetcode.com/problems/longest-palindromic-substring/
	// Idea: Expand around center
	// Alternative: This can be solved using DP with same complexity.
	// Time: O(n^2)
	public String longestPalindrome(String s) {
        if (s == null || s.length() < 1)
            return "";

        int start = 0, end = 0;
        for (int i = 0; i < s.length(); i++) {
            int len1 = expandAroundCenter(s, i, i);			// aba
            int len2 = expandAroundCenter(s, i, i + 1);		// abba
            int maxLen = Math.max(len1, len2);
            if (maxLen > end - start) {
                start = i - (maxLen - 1) / 2;
                end = i + maxLen / 2;
            }
        }

        return s.substring(start, end + 1);
    }

    private int expandAroundCenter(String s, int l, int r) {
        while (l >= 0 && r < s.length()) {
			if (s.charAt(l) != s.charAt(r))
				break;
            l--;
            r++;
        }
        return r - l - 1;
    }

	// Idea: Start with l=1 and go till l=n.
	// If first char and last char matches, add 2 to the lenth and look into the substring except first and last chars
	// Track the max len and substring which is polindrom when max gets updated
	// aba = 2+len(b) = 3
	//
	// Time: O(n^2)
	//
	// 	T[i][j] = T[i+1][j-1] if str[i] == str[j]
	//
	//
	//	i\j	b	a	b	a	d
	//	b	t	f	t	f	f
	//	a		t	f	t	f
	//	b			t	f	f
	//	a				t	f
	//	d					t
	//
	//	max = 3
	//
	// Time: O(n^2)
    // Space: O(n*n)
	public String longestPalindrome2(String s) {
		int n = s.length();
		boolean[][] dp = new boolean[n][n];

		int maxLen = 0;
		String polindrom = "";
		for(int l=0; l<n; l++) {
			for(int i=0; i<n-l; i++) {
				int j = i+l;
				if (i == j) {
					dp[i][j] = true;						// handles "b"
				} else if (s.charAt(i) == s.charAt(j)) {
					if(l == 1) {							// handles "bb"
						dp[i][j] = true;
					}else {
						dp[i][j] = dp[i + 1][j - 1];
					}
				}

				if(dp[i][j] && j-i+1 > maxLen) {
					maxLen = j-i+1;
					polindrom = s.substring(i, j+1);
				}
			}
		}
		return polindrom;
	}
}

File: dsa/concepts/18_strings/python/l5_longest_palindromic_substring.py
# Problem: https://leetcode.com/problems/longest-palindromic-substring/
# Idea: Expand around center
# Alternative: This can be solved using DP with same complexity.
# Time: O(n^2)
def longest_palindrome(s):
    if s is None or len(s) < 1:
        return ""

    start = 0
    end = 0

    for i in range(len(s)):
        len1 = expand_around_center(s, i, i)       # aba
        len2 = expand_around_center(s, i, i + 1)   # abba
        max_len = max(len1, len2)

        if max_len > end - start:
            start = i - (max_len - 1) // 2
            end = i + max_len // 2

    return s[start:end + 1]


def expand_around_center(s, l, r):
    while l >= 0 and r < len(s):
        if s[l] != s[r]:
            break
        l -= 1
        r += 1
    return r - l - 1


# Idea: Start with l=1 and go till l=n.
# If first char and last char matches, add 2 to the length and look into the substring
# except first and last chars
# Track the max len and substring which is polindrom when max gets updated
# aba = 2+len(b) = 3
#
# Time: O(n^2)
#
#   T[i][j] = T[i+1][j-1] if str[i] == str[j]
#
# Time: O(n^2)
# Space: O(n*n)
def longest_palindrome2(s):
    n = len(s)
    dp = [[False] * n for _ in range(n)]

    max_len = 0
    palindrome = ""

    for l in range(n):
        for i in range(n - l):
            j = i + l

            if i == j:
                dp[i][j] = True                     # handles "b"
            elif s[i] == s[j]:
                if l == 1:                          # handles "bb"
                    dp[i][j] = True
                else:
                    dp[i][j] = dp[i + 1][j - 1]

            if dp[i][j] and (j - i + 1) > max_len:
                max_len = j - i + 1
                palindrome = s[i:j + 1]

    return palindrome


# ------------------ Test (Same as Java main) ------------------

s = "babad"
print(longest_palindrome(s))
print(longest_palindrome2(s))
